pytheus/registry.py

```diff
@@ -0,0 +1,63 @@
+from threading import Lock
+from typing import Iterable, Protocol
+
+
+class Collector(Protocol):
+    name: str
+
+    def collect(self) -> Iterable:
+        pass
+
+
+class Registry(Protocol):
+    def register(self, collector: Collector):
+        pass
+
+    def unregister(self, collector: Collector):
+        pass
+
+    def collect(self) -> Iterable:
+        pass
+
+
+class CollectorRegistry:
+    def __init__(self, prefix: str = None) -> None:
+        self._lock = Lock()
+        self._prefix = prefix
+        self._collectors: dict[str, Collector] = {}
+
+    def register(self, collector: Collector) -> None:
+        with self._lock:
+            if collector.name in self._collectors:
```

Comment:
should we raise an error if this happens?
or maybe just add a warning log as we correctly handle the case, possibly for now just a TODO comment

Reply:
We shouldn't crash the entire application if this happens, so I'd advise for logging this as well.

The level of the log depends on some factors:
- `info`: This is definitely allowed and possible, we just log the event
- `warning`: This is unlikely and something to watch for, but we deal with it as something that could happen in edge cases
- `error`: This should never happen according to the code, this is a bug

Reply:
I vote for `warning` as well, because you should check it, but it will still work correctly as expected.
You cannot create two `MetricCollector` with the same name, that would be the same as having two different metrics with the same name that is not allowed in prometheus.

I mentioned the crashing behaviour just because it's what other projects are doing, but I strongly agree on just logging it.

Reply:
Crashing is also not bad idea because I guess this is part of the intinial setup and it's fine to crash before starting the service. For now I go with `warning` 👍

Reply:
yeah and no, in some crazy situation someone might add dynamically a metric later (why would you do that?), and it would automatically be registered and if it has the same name it would crash. But this is really far fetched sorry 😅